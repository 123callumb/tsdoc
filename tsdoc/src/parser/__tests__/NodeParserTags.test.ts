import { TestHelpers } from './TestHelpers';

test('00 Block tags: positive examples', () => {
  TestHelpers.parseAndMatchSnapshot([
    '/**',
    ' * @one ',
    ' * @two',
    ' */'
  ].join('\n'));
});

test('01 Block tags: negative examples', () => {
  TestHelpers.parseAndMatchSnapshot([
    '/**',
    ' * @ one ',
    ' * +@two ',
    ' * @two+ ',
    ' */'
  ].join('\n'));
});

test('02 Inline tags: simple, positive', () => {
  TestHelpers.parseAndMatchSnapshot([
    '/**',
    ' * {@one} ',
    ' * {@two } ',
    ' * {@three}{@four} ',
    ' * {@five ',
    ' *   } ',
    ' */'
  ].join('\n'));
});

test('03 Inline tags: simple, negative', () => {
  TestHelpers.parseAndMatchSnapshot([
    '/**',
    ' * {@ one} ',
    ' * { @two} ',
    ' * {@three',
    ' */'
  ].join('\n'));
});

test('04 Inline tags: complex, positive', () => {
  TestHelpers.parseAndMatchSnapshot([
    '/**',
    ' * {@one some content}',
    ' * {@two multi',
    ' * line}',
    ' */'
  ].join('\n'));
  TestHelpers.parseAndMatchSnapshot([
    '/**',
    ' * {@three @taglike}',
    ' */'
  ].join('\n'));
});

test('05 Inline tags: escaping, positive', () => {
  TestHelpers.parseAndMatchSnapshot([
    '/**',
    ' * {@one left \\{ right \\} backslash \\\\ }',
    ' */'
  ].join('\n'));
});

test('06 Inline tags: escaping, negative', () => {
  TestHelpers.parseAndMatchSnapshot([
    '/**',
    ' * {@one curly\\}',
    ' */'
  ].join('\n'));
  TestHelpers.parseAndMatchSnapshot([
    '/**',
    ' * {@two curly{}}',
    ' */'
  ].join('\n'));
  TestHelpers.parseAndMatchSnapshot([
    '/**',
    ' * three: }',
    ' */'
  ].join('\n'));
});
