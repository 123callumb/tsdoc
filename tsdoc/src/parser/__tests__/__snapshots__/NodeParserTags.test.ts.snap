// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`00 Block tags: positive examples 1`] = `
Object {
  "buffer": "/**[n] * @one [n] * @two[n] */",
  "lines": Array [
    "@one",
    "@two",
  ],
  "rootNode": Object {
    "kind": "Section",
    "nodes": Array [
      Object {
        "kind": "BlockTag",
        "nodePrefix": "@one",
      },
      Object {
        "kind": "SoftBreak",
        "nodePrefix": "[n]",
      },
      Object {
        "kind": "BlockTag",
        "nodePrefix": "@two",
      },
      Object {
        "kind": "SoftBreak",
        "nodePrefix": "[n]",
      },
    ],
  },
}
`;

exports[`01 Block tags: negative examples 1`] = `
Object {
  "buffer": "/**[n] * @ one [n] * +@two [n] * @two+ [n] */",
  "lines": Array [
    "@ one",
    "+@two",
    "@two+",
  ],
  "rootNode": Object {
    "kind": "Section",
    "nodes": Array [
      Object {
        "errorLocation": "@",
        "errorMessage": "Expecting a TSDoc tag name after the [q]@[q] character (or use a backslash to escape this character)",
        "kind": "ErrorText",
        "nodePrefix": "@",
      },
      Object {
        "kind": "PlainText",
        "nodePrefix": " one",
      },
      Object {
        "kind": "SoftBreak",
        "nodePrefix": "[n]",
      },
      Object {
        "kind": "PlainText",
        "nodePrefix": "+",
      },
      Object {
        "errorLocation": "@",
        "errorLocationPrecedingToken": "+",
        "errorMessage": "A TSDoc tag must be preceded by whitespace",
        "kind": "ErrorText",
        "nodePrefix": "@",
      },
      Object {
        "kind": "PlainText",
        "nodePrefix": "two",
      },
      Object {
        "kind": "SoftBreak",
        "nodePrefix": "[n]",
      },
      Object {
        "errorLocation": "@",
        "errorLocationPrecedingToken": "
",
        "errorMessage": "A TSDoc tag must be followed by whitespace",
        "kind": "ErrorText",
        "nodePrefix": "@",
      },
      Object {
        "kind": "PlainText",
        "nodePrefix": "two+",
      },
      Object {
        "kind": "SoftBreak",
        "nodePrefix": "[n]",
      },
    ],
  },
}
`;

exports[`02 Inline tags: simple, positive 1`] = `
Object {
  "buffer": "/**[n] * {@one} [n] * {@two } [n] * {@three}{@four} [n] * {@five [n] *   } [n] */",
  "lines": Array [
    "{@one}",
    "{@two }",
    "{@three}{@four}",
    "{@five",
    "  }",
  ],
  "rootNode": Object {
    "kind": "Section",
    "nodes": Array [
      Object {
        "kind": "InlineTag",
        "nodes": Array [
          Object {
            "kind": "Particle",
            "nodePrefix": "{",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": "@one",
          },
          Object {
            "kind": "Particle",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": "}",
          },
        ],
      },
      Object {
        "kind": "SoftBreak",
        "nodePrefix": "[n]",
      },
      Object {
        "kind": "InlineTag",
        "nodes": Array [
          Object {
            "kind": "Particle",
            "nodePrefix": "{",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": "@two",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": " ",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": "}",
          },
        ],
      },
      Object {
        "kind": "SoftBreak",
        "nodePrefix": "[n]",
      },
      Object {
        "kind": "InlineTag",
        "nodes": Array [
          Object {
            "kind": "Particle",
            "nodePrefix": "{",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": "@three",
          },
          Object {
            "kind": "Particle",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": "}",
          },
        ],
      },
      Object {
        "kind": "InlineTag",
        "nodes": Array [
          Object {
            "kind": "Particle",
            "nodePrefix": "{",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": "@four",
          },
          Object {
            "kind": "Particle",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": "}",
          },
        ],
      },
      Object {
        "kind": "SoftBreak",
        "nodePrefix": "[n]",
      },
      Object {
        "kind": "InlineTag",
        "nodes": Array [
          Object {
            "kind": "Particle",
            "nodePrefix": "{",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": "@five",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": "[n]  ",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": "}",
          },
        ],
      },
      Object {
        "kind": "SoftBreak",
        "nodePrefix": "[n]",
      },
    ],
  },
}
`;

exports[`03 Inline tags: simple, negative 1`] = `
Object {
  "buffer": "/**[n] * {@ one} [n] * {@two~} [n] * { @three} [n] * {@four[n] */",
  "lines": Array [
    "{@ one}",
    "{@two~}",
    "{ @three}",
    "{@four",
  ],
  "rootNode": Object {
    "kind": "Section",
    "nodes": Array [
      Object {
        "errorLocation": "@",
        "errorLocationPrecedingToken": "{",
        "errorMessage": "Expecting a TSDoc inline tag name after the [q]{@[q] characters",
        "kind": "ErrorText",
        "nodePrefix": "{@",
      },
      Object {
        "kind": "PlainText",
        "nodePrefix": " one",
      },
      Object {
        "errorLocation": "}",
        "errorLocationPrecedingToken": "one",
        "errorMessage": "The [q]}[q] character should be escaped using a backslash to avoid confusion with a TSDoc inline tag",
        "kind": "ErrorText",
        "nodePrefix": "}",
      },
      Object {
        "kind": "SoftBreak",
        "nodePrefix": "[n]",
      },
      Object {
        "errorLocation": "~",
        "errorLocationPrecedingToken": "two",
        "errorMessage": "Expecting a space after the TSDoc inline tag name",
        "kind": "ErrorText",
        "nodePrefix": "{@",
      },
      Object {
        "kind": "PlainText",
        "nodePrefix": "two~",
      },
      Object {
        "errorLocation": "}",
        "errorLocationPrecedingToken": "~",
        "errorMessage": "The [q]}[q] character should be escaped using a backslash to avoid confusion with a TSDoc inline tag",
        "kind": "ErrorText",
        "nodePrefix": "}",
      },
      Object {
        "kind": "SoftBreak",
        "nodePrefix": "[n]",
      },
      Object {
        "errorLocation": "{",
        "errorLocationPrecedingToken": "
",
        "errorMessage": "Expecting a TSDoc tag starting with [q]{@[q]",
        "kind": "ErrorText",
        "nodePrefix": "{",
      },
      Object {
        "kind": "PlainText",
        "nodePrefix": " ",
      },
      Object {
        "errorLocation": "@",
        "errorLocationPrecedingToken": " ",
        "errorMessage": "A TSDoc tag must be followed by whitespace",
        "kind": "ErrorText",
        "nodePrefix": "@",
      },
      Object {
        "kind": "PlainText",
        "nodePrefix": "three",
      },
      Object {
        "errorLocation": "}",
        "errorLocationPrecedingToken": "three",
        "errorMessage": "The [q]}[q] character should be escaped using a backslash to avoid confusion with a TSDoc inline tag",
        "kind": "ErrorText",
        "nodePrefix": "}",
      },
      Object {
        "kind": "SoftBreak",
        "nodePrefix": "[n]",
      },
      Object {
        "errorLocation": "{@",
        "errorLocationPrecedingToken": "
",
        "errorMessage": "The TSDoc inline tag name is missing its closing [q]}[q]",
        "kind": "ErrorText",
        "nodePrefix": "{@",
      },
      Object {
        "kind": "PlainText",
        "nodePrefix": "four",
      },
      Object {
        "kind": "SoftBreak",
        "nodePrefix": "[n]",
      },
    ],
  },
}
`;

exports[`04 Inline tags: complex, positive 1`] = `
Object {
  "buffer": "/**[n] * {@one some content}[n] * {@two multi[n] * line}[n] */",
  "lines": Array [
    "{@one some content}",
    "{@two multi",
    "line}",
  ],
  "rootNode": Object {
    "kind": "Section",
    "nodes": Array [
      Object {
        "kind": "InlineTag",
        "nodes": Array [
          Object {
            "kind": "Particle",
            "nodePrefix": "{",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": "@one",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": " some content",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": "}",
          },
        ],
      },
      Object {
        "kind": "SoftBreak",
        "nodePrefix": "[n]",
      },
      Object {
        "kind": "InlineTag",
        "nodes": Array [
          Object {
            "kind": "Particle",
            "nodePrefix": "{",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": "@two",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": " multi[n]line",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": "}",
          },
        ],
      },
      Object {
        "kind": "SoftBreak",
        "nodePrefix": "[n]",
      },
    ],
  },
}
`;

exports[`04 Inline tags: complex, positive 2`] = `
Object {
  "buffer": "/**[n] * {@three @taglike}[n] */",
  "lines": Array [
    "{@three @taglike}",
  ],
  "rootNode": Object {
    "kind": "Section",
    "nodes": Array [
      Object {
        "kind": "InlineTag",
        "nodes": Array [
          Object {
            "kind": "Particle",
            "nodePrefix": "{",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": "@three",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": " @taglike",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": "}",
          },
        ],
      },
      Object {
        "kind": "SoftBreak",
        "nodePrefix": "[n]",
      },
    ],
  },
}
`;

exports[`05 Inline tags: escaping, positive 1`] = `
Object {
  "buffer": "/**[n] * {@one left [b]{ right [b]} backslash [b][b] }[n] */",
  "lines": Array [
    "{@one left [b]{ right [b]} backslash [b][b] }",
  ],
  "rootNode": Object {
    "kind": "Section",
    "nodes": Array [
      Object {
        "kind": "InlineTag",
        "nodes": Array [
          Object {
            "kind": "Particle",
            "nodePrefix": "{",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": "@one",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": " left [b]{ right [b]} backslash [b][b] ",
          },
          Object {
            "kind": "Particle",
            "nodePrefix": "}",
          },
        ],
      },
      Object {
        "kind": "SoftBreak",
        "nodePrefix": "[n]",
      },
    ],
  },
}
`;

exports[`06 Inline tags: escaping, negative 1`] = `
Object {
  "buffer": "/**[n] * {@one curly[b]}[n] */",
  "lines": Array [
    "{@one curly[b]}",
  ],
  "rootNode": Object {
    "kind": "Section",
    "nodes": Array [
      Object {
        "errorLocation": "{@",
        "errorMessage": "The TSDoc inline tag name is missing its closing [q]}[q]",
        "kind": "ErrorText",
        "nodePrefix": "{@",
      },
      Object {
        "kind": "PlainText",
        "nodePrefix": "one curly",
      },
      Object {
        "kind": "EscapedText",
        "nodePrefix": "[b]}",
      },
      Object {
        "kind": "SoftBreak",
        "nodePrefix": "[n]",
      },
    ],
  },
}
`;

exports[`06 Inline tags: escaping, negative 2`] = `
Object {
  "buffer": "/**[n] * {@two curly{}}[n] */",
  "lines": Array [
    "{@two curly{}}",
  ],
  "rootNode": Object {
    "kind": "Section",
    "nodes": Array [
      Object {
        "errorLocation": "{",
        "errorLocationPrecedingToken": "curly",
        "errorMessage": "The [q]{[q] character must be escaped with a backslash when used inside a TSDoc inline tag",
        "kind": "ErrorText",
        "nodePrefix": "{@",
      },
      Object {
        "kind": "PlainText",
        "nodePrefix": "two curly",
      },
      Object {
        "errorLocation": "{",
        "errorLocationPrecedingToken": "curly",
        "errorMessage": "Expecting a TSDoc tag starting with [q]{@[q]",
        "kind": "ErrorText",
        "nodePrefix": "{",
      },
      Object {
        "errorLocation": "}",
        "errorLocationPrecedingToken": "{",
        "errorMessage": "The [q]}[q] character should be escaped using a backslash to avoid confusion with a TSDoc inline tag",
        "kind": "ErrorText",
        "nodePrefix": "}",
      },
      Object {
        "errorLocation": "}",
        "errorLocationPrecedingToken": "}",
        "errorMessage": "The [q]}[q] character should be escaped using a backslash to avoid confusion with a TSDoc inline tag",
        "kind": "ErrorText",
        "nodePrefix": "}",
      },
      Object {
        "kind": "SoftBreak",
        "nodePrefix": "[n]",
      },
    ],
  },
}
`;

exports[`06 Inline tags: escaping, negative 3`] = `
Object {
  "buffer": "/**[n] * three: }[n] */",
  "lines": Array [
    "three: }",
  ],
  "rootNode": Object {
    "kind": "Section",
    "nodes": Array [
      Object {
        "kind": "PlainText",
        "nodePrefix": "three: ",
      },
      Object {
        "errorLocation": "}",
        "errorLocationPrecedingToken": " ",
        "errorMessage": "The [q]}[q] character should be escaped using a backslash to avoid confusion with a TSDoc inline tag",
        "kind": "ErrorText",
        "nodePrefix": "}",
      },
      Object {
        "kind": "SoftBreak",
        "nodePrefix": "[n]",
      },
    ],
  },
}
`;
